src/python/pants/util/objects.py

```diff
@@ -13,6 +13,9 @@
 from pants.util.meta import AbstractClass
 
 
+class DatatypeHelperConstructionError(Exception): pass
```

Comment:
The mild style bias is towards nested exception types, this allows, for example:
```python
with self.assertRaises(datatype.DatatypeHelperConstructionError):
  ...
```

IE: no specific import for the exception type, just a focus on importing the nominal type.

Also, since there is no DataypeHelper noun floating around, `DatatypeConstructionError` would make more sense to me. In any case it seems like it should extend `TypeError` and if that is enough distinction for tests (I think it is), you could just drop the cutom exception type altogether!

Reply:
I've noted the description of nested exception types for future reference, thanks (the imports part makes sense to me). I have removed the `Helper` from the exception type name, made it subclass `TypeError`, then caught `TypeError` in the test case for it.

Reply:
Well, if the test is checking for TypeError, there is no remaining point to a custom exception type. It's only useful to have one if someone can benefit and discriminate on the difference from a built in exception type.

Reply:
Ok, that makes sense. I have removed the custom exception type (and changed the `:raises:` annotation) in f7c2565.